### PR TITLE
Mc knight 42/test event codes

### LINKED ISSUE #4329: Clean up types.py for logging
validate `code` values are unique for each event in `core/dbt/events/types.py`
    - ex: E001 occurs twice

validate event is called only once in `dump_callable`

### PR DESCRIPTION
**Pull request by McKnight-42**
resolves #4329 

### Description
Created a test to check for duplicate event code values.

### Checklist

- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have updated the `CHANGELOG.md` and added information about my change

### CODE REVIEW

File: test/unit/test_events.py [hunk 1]
```diff
@@ -45,3 +46,32 @@ def test_formatting(self):
         # were no args passed after the initial message
         event = AdapterEventDebug(name="dbt_tests", base_msg="boop{x}boop", args=())
         self.assertTrue("boop{x}boop" in event.message())
+
+class TestEventCodes(TestCase):
+
+    def get_all_subclasses(self,cls):
```
Comment (nathaniel-may) [inline test/unit/test_events.py:52]: minor formatting thing
```suggestion
    def get_all_subclasses(self, cls):
```

File: test/unit/test_events.py [hunk 2]
```diff
@@ -45,3 +46,32 @@ def test_formatting(self):
...
+        all_subclasses = []
+        for subclass in cls.__subclasses__():
+            all_subclasses.append(subclass)
+            all_subclasses.extend(self.get_all_subclasses(subclass))
+        return set(all_subclasses)
+
+
+    def setUp(self):
+        pass
```
Comment (nathaniel-may) [inline test/unit/test_events.py:61]: let's move this up top so it's the first method in the class

File: test/unit/test_events.py [hunk 3]
```diff
@@ -45,3 +46,32 @@ def test_formatting(self):
...
+
+    # this interface is documented for adapter maintainers to plug into
+    # so we should test that it at the very least doesn't explode.
```
Comment (nathaniel-may) [inline test/unit/test_events.py:64]: This comment is probably a copy paste error. Let's write a new one that describes what this tests for and why it's important.

File: test/unit/test_events.py [hunk 4]
```diff
@@ -45,3 +46,32 @@ def test_formatting(self):
...
+    def test_event_codes(self):
+        # all_concrete = # all subclasses (any depth) of Event
```
Comment (nathaniel-may) [inline test/unit/test_events.py:66]: we can remove this comment

File: test/unit/test_events.py [hunk 5]
```diff
@@ -45,3 +46,32 @@ def test_formatting(self):
...
+        all_concrete = self.get_all_subclasses(Event)
+        all_codes = set()
+
+        for event in all_concrete:
+            if not inspect.isabstract(event):
+                # must be in the form 1 capital letter, 3 digits
+                self.assertTrue('^[A-Z][0-9]{3}', event.code)
+                # cannot have been used already
+                self.assertFalse(event.code in all_codes,f'{event.code}')
```
Comment (nathaniel-may) [inline test/unit/test_events.py:75]: minor formatting, and helper string improvement
```suggestion
                self.assertFalse(event.code in all_codes, f'Code {event.code} is assigned more than once. Check types.py for duplicates.')
```

**Review by nathaniel-may — CHANGES_REQUESTED** [review]

File: test/unit/test_events.py (hunk 5 above)
Comment (McKnight-42) [reply test/unit/test_events.py]: Thank you for all the catches, made changes and repushed.

File: test/unit/test_events.py [hunk 6]
```diff
@@ -1,8 +1,9 @@
 from dbt.events import AdapterLogger
 from dbt.events.types import AdapterEventDebug
+from dbt.events.base_types import Cli,File,Event
```
Comment (nathaniel-may) [inline test/unit/test_events.py:3]: Just the commas here too
```suggestion
from dbt.events.base_types import Cli, File, Event
```

File: test/unit/test_events.py [hunk 7]
```diff
@@ -45,3 +46,29 @@ def test_formatting(self):
         # were no args passed after the initial message
         event = AdapterEventDebug(name="dbt_tests", base_msg="boop{x}boop", args=())
         self.assertTrue("boop{x}boop" in event.message())
+
+class TestEventCodes(TestCase):
+
+    def setUp(self):
+        pass
+
+    def get_all_subclasses(self, cls):
```
Comment (nathaniel-may) [inline test/unit/test_events.py:55]: don't forget to document this

File: test/unit/test_events.py [hunk 8]
```diff
@@ -45,3 +46,29 @@ def test_formatting(self):
...
+        all_subclasses = []
+        for subclass in cls.__subclasses__():
+            all_subclasses.append(subclass)
+            all_subclasses.extend(self.get_all_subclasses(subclass))
+        return set(all_subclasses)
+
+    # checks to see if event codes are duplicated to keep codes singluar and clear.
```
Comment (nathaniel-may) [inline test/unit/test_events.py:62]: it also checks the code format with `self.assertTrue('^[A-Z][0-9]{3}', event.code)`

**Review by nathaniel-may — CHANGES_REQUESTED** [review]
Looking good! Just some more minor stuff.

File: test/unit/test_events.py (hunk 6 above)
Comment (McKnight-42) [reply test/unit/test_events.py:3]: Removed unused imports

File: test/unit/test_events.py (hunk 7 above)
Comment (McKnight-42) [reply test/unit/test_events.py:55]: added documentation for get_all_subclassses, and removed setUp method as its unused currently.

File: test/unit/test_events.py (hunk 8 above)
Comment (McKnight-42) [reply test/unit/test_events.py:62]: added more clarifying comments.

**Review by nathaniel-may — APPROVED** [review]
Great work!

**Pull request merged**